test/lib/PlacesTestUtils.js

```diff
@@ -69,7 +118,28 @@ const PlacesTestUtils = Object.freeze({
         }
       );
     });
-    return (yield promise);
+
+    yield notifPromise;
+    yield insertPromise;
+
+    // poll every 10ms until history items appear on disk
+    yield waitUntil(function*() {
```

Comment:
Re: efficiency, that assumption is incorrect. I encourage you to watch: 
https://www.youtube.com/watch?v=65-RbBwZQdU

(I'll try to find the exact time where he shows that the VM can optimize declarative calls, like `.filter()` better than imperative for loops)